### PR TITLE
Some updates to make the debian packageing work.

### PR DESCRIPTION
**Pull request by Kunzol**
## Updates to make debuild work again.

Because the directory was moved to the subdir dist, it is needed
to copy the directory to the top directory or set a link
before running debuild.

As usual the command for building the debian package is:
```
debuild -b -us -uc
```
To make it configurable some environment variables can be set.
Due to the fact, that debuild filters out some variables, all the
configuration variables need to be prefixed by DEB_

First the changelog file is updated with the current version. 
To configure the fullname and email available in the changelog
message set the environment variables:
```
export DEBFULLNAME="Your Name"
export DEB_FULLNAME=${DEBFULLNAME}
export EMAIL=<your email>
```
options for mconfig, by setting the environment variables:
`DEB_NOSUID=1 `   adds --without-suid
`DEB_NONETWORK=1` adds --without-network
`DEB_NOSECCOMP=1` adds --without-seccomp
`DEB_NOALL=1`     adds all of the above

To select a specific profile for mconfig:
```
export DEB_SC_PROFILE=release-stripped
```
or
```
export DEB_SC_PROFILE=debug
```
The builddir can be specified with this:
```
export DEB_SC_BUILDDIR=builddir
```
Currently no other config options can be set via environment,
but it would be easy to add them if needed.

I used as much as possible from the existing rules file for this
first step.

In the control file I added a build dependency to golang and changed
the architecture to amd64.

I do not expect that it is already perfect, but I hope this helps others in need 
for a debian package of singularity-container.

### CODE REVIEW

**Comment by DrDaveD** [comment]
Cool, thanks.  I think it would be good to have the detailed info about building and configuration options (in the description of this PR) put into a file in the dist/debian directory and referred to from INSTALL.md.

**Review by DrDaveD — CHANGES_REQUESTED** [review]
Thanks for the description file.  Please add a reference to it in INSTALL.md too so people can find it.  I think a new short section at the end is appropriate.

File: INSTALL.md [hunk 1]
```diff
@@ -161,3 +161,9 @@ $ make -C builddir rpm RPMPREFIX=/usr/local
 
 For more information on installing/updating/uninstalling the RPM, check out our 
 [admin docs](https://www.sylabs.io/guides/3.0/admin-guide/admin_quickstart.html).
+
+## Debian Package
+
+Additional information on [how to build a Debian package](dist/debian/DEBIAN_PACKAGE.md) can be 
+found in `dist/debian`.
```
Comment (DrDaveD) [inline INSTALL.md]:
```suggestion
Additional information on how to build a Debian package can be found in [dist/debian/DEBIAN_PACKAGE.md](dist/debian/DEBIAN_PACKAGE.md).
```

File: dist/debian/control [hunk 2]
```diff
@@ -15,14 +15,15 @@ Build-Depends:
...
+ golang-go (>= 2:1.13)
 Standards-Version: 3.9.8
 Homepage: http://gmkurtzer.github.io/singularity
 Vcs-Git: https://github.com/hpcng/singularity.git
 Vcs-Browser: https://github.com/hpcng/singularity
 
 # "singularity" is a packaged game (but the contents don't clash)
 Package: singularity-container
-Architecture: any
+Architecture: amd64
```
Comment (DrDaveD) [inline dist/debian/control]: What happens if you don't make this change?  I still have an i386 debian machine and it complains about this.

**Review by DrDaveD — CHANGES_REQUESTED** [review]

**Comment by DrDaveD** [comment]
I tested the debuild command on a branch of this pr on my Debian stable (mostly, with some packages from testing) machine and got the following output: 
```
$ debuild --build=binary --no-sign --lintian-opts --display-info --show-overrides 
 dpkg-buildpackage -us -uc -ui --build=binary --build=binary
dpkg-buildpackage: info: source package singularity-container
dpkg-buildpackage: info: source version 2.4.2-1
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Gregory M. Kurtzer <gmkurtzer@gmail.com>
 dpkg-source --before-build .
dpkg-buildpackage: info: host architecture i386
 fakeroot debian/rules clean
dh clean  --with=autoreconf
   dh_auto_clean
	make -j1 distclean
make[1]: Entering directory '/space/home/dwd/work/singularity'
make[1]: *** No rule to make target 'Makefile.am', needed by 'Makefile.in'.  Stop.
make[1]: Leaving directory '/space/home/dwd/work/singularity'
dh_auto_clean: make -j1 distclean returned exit code 2
make: *** [debian/rules:55: clean] Error 2
dpkg-buildpackage: error: fakeroot debian/rules clean subprocess returned exit status 2
debuild: fatal error at line 1182:
dpkg-buildpackage -us -uc -ui --build=binary --build=binary failed
```
Do you have any suggestions for me?

Perhaps I should start simpler and test the build from source instructions first, since I haven't built singularity on this machine for a very long time (pre-go days).

**Comment by DrDaveD** [comment]
I would also like to see this become part of the debian CI test, but we can do that in a separate PR.

File: dist/debian/control (hunk 2 above)
Comment (Kunzol) [reply dist/debian/control]: I did not expect, that a tool designed for HPC like singularity is useful on 32bit. Maybe I can find a 32bit Debian to test it.

**Comment by DrDaveD** [comment]
I successfully built from source on the machine using INSTALL.md instructions, so I'm still stuck on the debuild.

Note that currently INSTALL.md conflicts with the master branch, that will need to be resolved before merging.

It would also be great if someone could manage to update and maintain the [singularity-container debian package](https://tracker.debian.org/pkg/singularity-container).  It was removed from Debian testing last year.

File: dist/debian/control (hunk 2 above)
Comment (DrDaveD) [reply dist/debian/control]: It's probably not very useful, it's just that it's the machine I have handy!  You could leave that testing to me, but what happens on your machine if you leave it as "any"?  debuild appears to figure out what architecture to put in there.

**Comment by Kunzol** [comment]
Just for completeness how the build should look like: 

```
$ ln -s dist/debian .
$ debuild --build=binary --no-sign --lintian-opts --display-info --show-overrides
 dpkg-buildpackage -us -uc -ui --build=binary --build=binary
dpkg-buildpackage: info: source package singularity-container
dpkg-buildpackage: info: source version 2.4.2-1
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Gregory M. Kurtzer <gmkurtzer@gmail.com>
 dpkg-source --before-build .
dpkg-buildpackage: info: host architecture amd64
 fakeroot debian/rules clean
dh clean  --with=autoreconf
   debian/rules override_dh_clean
make[1]: Entering directory 'singularity-kunzol'
make[1]: Leaving directory 'singularity-kunzol'
 debian/rules build
dh build  --with=autoreconf
   dh_update_autotools_config
   dh_autoreconf
   debian/rules override_dh_auto_configure
make[1]: Entering directory 'singularity-kunzol'
debian/rules:59: "Setting new version in debian changelog: 3.8.0-rc.2+12-g728f75793"
Configuring for project `singularity' with languages: C, Golang
=> running pre-basechecks project specific checks ...
=> running base system checks ...
 checking: host C compiler... cc
 checking: host C++ compiler... c++
 checking: host Go compiler (at least version 1.13)... /usr/bin/go
 checking: host C compiler option -Wall... yes
 checking: host C compiler option -Werror... yes
 checking: host C compiler option -Wfatal-errors... yes
 checking: host C compiler option -Wno-unknown-warning-option... yes
 checking: host C compiler option -Wstrict-prototypes... yes
 checking: host C compiler option -Wpointer-arith... yes
 checking: host C compiler option -Wbad-function-cast... yes
 checking: host C compiler option -Woverlength-strings... yes
 checking: host C compiler option -Wframe-larger-than=2047... yes
 checking: host C compiler option -Wno-sign-compare... yes
 checking: host C compiler option -Wclobbered... yes
 checking: host C compiler option -Wempty-body... yes
 checking: host C compiler option -Wmissing-parameter-type... yes
 checking: host C compiler option -Wtype-limits... yes
 checking: host C compiler option -Wunused-parameter... yes
 checking: host C compiler option -Wunused-but-set-parameter... yes
 checking: host C compiler option -Wno-discarded-qualifiers... yes
 checking: host C compiler option -Wno-incompatible-pointer-types... yes
 checking: host C compiler option -pipe... yes
 checking: host C compiler option -fmessage-length=0... yes
 checking: host C compiler option -fPIC... yes
 checking: host `ar' path... ar
 checking: host `ld' path... ld
 checking: host `ranlib' path... ranlib
 checking: host `objcopy' path... objcopy
 checking: target C compiler... cc
 checking: target C++ compiler... c++
 checking: target `ar' path... ar
 checking: target `ld' path... ld
 checking: target `ranlib' path... ranlib
 checking: target `objcopy' path... objcopy
 checking: host compiles static binaries... yes
 checking: target compiles static binaries... yes
 checking: host os type... unix
 checking: host architecture... x86_64
 checking: target architecture... x86_64
 checking: host architecture word size... 64
 checking: target architecture word size... 64
 checking: project version... 3.8.0-rc.2+12-g728f75793-dirty
 checking: project short version... 3.8.0-rc.2+12-g728f75793-dirty
=> running post-basechecks project specific checks ...
 checking: namespace: CLONE_NEWPID... yes
 checking: namespace: CLONE_FS... yes
 checking: namespace: CLONE_NEWNS... yes
 checking: namespace: CLONE_NEWUSER... yes
 checking: namespace: CLONE_NEWIPC... yes
 checking: namespace: CLONE_NEWNET... yes
 checking: namespace: CLONE_NEWUTS... yes
 checking: namespace: CLONE_NEWCGROUP... yes
 checking: feature: NO_NEW_PRIVS... yes
 checking: feature: MS_SLAVE... yes
 checking: feature: MS_REC... yes
 checking: feature: MS_PRIVATE... yes
 checking: user capabilities... yes
 checking: header linux/securebits.h... yes
 checking: header linux/capability.h... yes
 checking: libseccomp+headers... no
 checking: cryptsetup... no

unable to find the cryptsetup program, is the package cryptsetup-bin installed?

=> generating fragments ...
=> building Makefile ...
=> generating singularity.spec ...
=> project singularity setup with :
    - host arch: x86_64
    - host wordsize: 64-bit
    - host C compiler: cc
    - host Go compiler: /usr/bin/go
    - host system: unix
      ---
    - target arch: x86_64
    - target wordsize: 64-bit
    - target C compiler: cc
      ---
    - config profile: release
      ---
    - SUID install: yes
    - Network plugins: yes
      ---
    - verbose: no
      ---
    - cryptsetup: 
      ---
    - version: 3.8.0-rc.2+12-g728f75793-dirty
=> singularity-kunzol/builddir/Makefile ready, try:
   $ cd singularity-kunzol/builddir
   $ make
make[1]: Leaving directory 'singularity-kunzol'
   debian/rules override_dh_auto_build
make[1]: Entering directory 'singularity-kunzol'
        cd builddir && make -j30
make[2]: Entering directory 'singularity-kunzol/builddir'
 GEN GO DEP singularity-kunzol/builddir/starter.d
 GEN GO DEP singularity-kunzol/builddir/singularity.d
 GO singularity
    [+] GO_TAGS "containers_image_openpgp sylog oci_engine singularity_engine fakeroot_engine apparmor selinux"
 GEN etc/bash_completion.d/singularity
 GEN singularity.conf from /etc/singularity/singularity.conf
 CNI PLUGIN dhcp
 CNI PLUGIN host-local
 CNI PLUGIN static
 CNI PLUGIN bridge
 CNI PLUGIN host-device
 CNI PLUGIN ipvlan
 CNI PLUGIN loopback
 CNI PLUGIN macvlan
 CNI PLUGIN ptp
 CNI PLUGIN vlan
 CNI PLUGIN bandwidth
 CNI PLUGIN firewall
 CNI PLUGIN flannel
 CNI PLUGIN portmap
 CNI PLUGIN tuning
 GO clean -cache
 GO cmd/starter/c/starter
 GO cmd/starter/c/starter-suid
    [+] GO_TAGS "containers_image_openpgp sylog singularity_engine fakeroot_engine apparmor selinux"
 GEN singularity-kunzol/scripts/go-test
make[2]: Leaving directory 'singularity-kunzol/builddir'
make[1]: Leaving directory 'singularity-kunzol'
 fakeroot debian/rules binary
dh binary  --with=autoreconf
   dh_testroot
   dh_prep
   dh_installdirs
   debian/rules override_dh_auto_install
make[1]: Entering directory 'singularity-kunzol'
        cd builddir && make -j1 install DESTDIR=singularity-kunzol/debian/singularity-container AM_UPDATE_INFO_DIR=no
make[2]: Entering directory 'singularity-kunzol/builddir'
 INSTALL singularity-kunzol/debian/singularity-container/usr/bin/singularity
 INSTALL singularity-kunzol/debian/singularity-container/etc/bash_completion.d/singularity
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/singularity.conf
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/remote.yaml
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/dhcp
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/host-local
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/static
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/bridge
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/host-device
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/ipvlan
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/loopback
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/macvlan
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/ptp
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/vlan
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/bandwidth
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/firewall
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/flannel
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/portmap
 INSTALL CNI PLUGIN singularity-kunzol/debian/singularity-container/usr/libexec/singularity/cni/tuning
 INSTALL CNI CONFIGURATION FILES
 INSTALL singularity-kunzol/debian/singularity-container/usr/libexec/singularity/bin/starter
 INSTALL singularity-kunzol/debian/singularity-container/usr/bin/run-singularity
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/capability.json
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/ecl.toml
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/seccomp-profiles/default.json
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/nvliblist.conf
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/rocmliblist.conf
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/cgroups/cgroups.toml
 INSTALL singularity-kunzol/debian/singularity-container/etc/singularity/global-pgp-public
 INSTALL SUID singularity-kunzol/debian/singularity-container/usr/libexec/singularity/bin/starter-suid
 DONE
make[2]: Leaving directory 'singularity-kunzol/builddir'
make[1]: Leaving directory 'singularity-kunzol'
   dh_installdocs
   dh_installchangelogs
   dh_installexamples
   debian/rules override_dh_installman
make[1]: Entering directory 'singularity-kunzol'
: # Very sloppy man pages for now
make[1]: Leaving directory 'singularity-kunzol'
   dh_lintian
   dh_perl
   dh_link
   dh_strip_nondeterminism
   dh_compress
   debian/rules override_dh_fixperms
make[1]: Entering directory 'singularity-kunzol'
make[1]: Leaving directory 'singularity-kunzol'
   dh_missing
   dh_strip
   dh_makeshlibs
   dh_shlibdeps
   dh_installdeb
   dh_gencontrol
   dh_md5sums
   dh_builddeb
dpkg-deb: building package 'singularity-container-dbgsym' in '../singularity-container-dbgsym_3.8.0-rc.2+12-g728f75793_amd64.deb'.
dpkg-deb: building package 'singularity-container' in '../singularity-container_3.8.0-rc.2+12-g728f75793_amd64.deb'.
 dpkg-genbuildinfo --build=binary
 dpkg-genchanges --build=binary >../singularity-container_2.4.2-1_amd64.changes
dpkg-genchanges: info: binary-only upload (no source code included)
 dpkg-source --after-build .
dpkg-buildpackage: info: binary-only upload (no source included)
Now running lintian --display-info --show-overrides singularity-container_2.4.2-1_amd64.changes ...
I: singularity-container: spelling-error-in-binary usr/bin/singularity dont don't
I: singularity-container: spelling-error-in-binary usr/bin/singularity lenth length
I: singularity-container: spelling-error-in-binary usr/bin/singularity formated formatted
I: singularity-container: spelling-error-in-binary usr/bin/singularity sylog syslog
I: singularity-container: spelling-error-in-binary usr/bin/singularity thats that's
I: singularity-container: spelling-error-in-binary usr/bin/singularity lIk link
I: singularity-container: spelling-error-in-binary usr/bin/singularity allows to allows one to
I: singularity-container: spelling-error-in-binary usr/bin/singularity allow to allow one to
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/bin/starter sylog syslog
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/bin/starter-suid sylog syslog
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/bandwidth
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/cni/bridge promiscous promiscuous
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/cni/bridge Prefered Preferred
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/bridge
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/dhcp
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/firewall
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/flannel
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/cni/host-device Prefered Preferred
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/host-device
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/host-local
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/cni/ipvlan Prefered Preferred
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/ipvlan
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/loopback
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/cni/macvlan Prefered Preferred
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/macvlan
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/portmap
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/cni/ptp Prefered Preferred
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/ptp
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/static
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/tuning
I: singularity-container: spelling-error-in-binary usr/libexec/singularity/cni/vlan Prefered Preferred
I: singularity-container: hardening-no-bindnow usr/libexec/singularity/cni/vlan
N: Singularity requires root suid for operation
O: singularity-container: setuid-binary usr/libexec/singularity/bin/starter-suid 4755 root/root
I: singularity-container: spelling-error-in-manpage usr/share/man/man1/run-singularity.1.gz dont don't
I: singularity-container: spelling-error-in-manpage usr/share/man/man1/run-singularity.1.gz formated formatted
Finished running lintian.
```

**Comment by Kunzol** [comment]
> I tested the debuild command on a branch of this pr on my Debian stable (mostly, with some packages from testing) machine and got the following output:
> 
> ```
> $ debuild --build=binary --no-sign --lintian-opts --display-info --show-overrides 
>  dpkg-buildpackage -us -uc -ui --build=binary --build=binary
> dpkg-buildpackage: info: source package singularity-container
> dpkg-buildpackage: info: source version 2.4.2-1
> dpkg-buildpackage: info: source distribution unstable
> dpkg-buildpackage: info: source changed by Gregory M. Kurtzer <gmkurtzer@gmail.com>
>  dpkg-source --before-build .
> dpkg-buildpackage: info: host architecture i386
>  fakeroot debian/rules clean
> dh clean  --with=autoreconf
>    dh_auto_clean
> 	make -j1 distclean
> make[1]: Entering directory '/space/home/dwd/work/singularity'
> make[1]: *** No rule to make target 'Makefile.am', needed by 'Makefile.in'.  Stop.
> make[1]: Leaving directory '/space/home/dwd/work/singularity'
> dh_auto_clean: make -j1 distclean returned exit code 2
> make: *** [debian/rules:55: clean] Error 2
> dpkg-buildpackage: error: fakeroot debian/rules clean subprocess returned exit status 2
> debuild: fatal error at line 1182:
> dpkg-buildpackage -us -uc -ui --build=binary --build=binary failed
> ```
> 
> Do you have any suggestions for me?
> 
> Perhaps I should start simpler and test the build from source instructions first, since I haven't built singularity on this machine for a very long time (pre-go days).

Are you sure you use my branch?
Looks not like the `rules` file from my branch is used. 

Im my `rules` flle the `dh_clean` has an "override", which is not mentionied in your building test.

**Comment by DrDaveD** [comment]
I am in your branch, but I found the problem.  My working space wasn't clean.  I had a "Makefile" at the top level that didn't belong.

File: dist/debian/control (hunk 2 above)
Comment (Kunzol) [reply dist/debian/control]: By the way, the semantic of the Architecture line is not the build target.

A value of `all` means that it is an **architecture-independent** package, which can be installed on all architectures. For example a package which only contains bash-scripts. 

This is not true for singularity-container package. It is ether amd64 or i386, depending on the architecture the package is build. It would need two separate Debian packages, one for amd64 and one for i386 (which needs to be built on a i386 or the compiler has to produce i386 binaries.) But this is not the task of "debuild", this must be a configuration in the sourcecode (or mconfig) of singularity.

**Comment by Kunzol** [comment]
> 
> Note that currently INSTALL.md conflicts with the master branch, that will need to be resolved before merging.

I do not have the right to resolve the conflict in INSTALL.md. 

> 
> It would also be great if someone could manage to update and maintain the [singularity-container debian package](https://tracker.debian.org/pkg/singularity-container). It was removed from Debian testing last year.

This is beyond my knowledge and far more than just creating the "debian" directory for building simple Debian packages.

As you can see in my build example and the `lintian` output, you first have to resolve all the spelling errors in the singularity code.  Which is most probably the easiest task on the way to a Debian package in the Debian Repository.

Additionally I be leave that singularity is a project in heavy development and as the users are researchers (at universities?) they usually want more or less the newest version all the time.

**Comment by DrDaveD** [comment]
My build completed.  

Please mention that the target packages go into directory '..' from where you run the command.

Consider adding some files go .gitignore.  After the build 'git status' shows a number of extra files:
```
hanges not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   dist/debian/changelog
	modified:   dist/debian/control

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	debian
	dist/debian/.debhelper/
	dist/debian/files
	dist/debian/singularity-container.debhelper.log
	dist/debian/singularity-container.substvars
	dist/debian/singularity-container/
```
I don't know how many of those should be ignored or how many should be cleaned out.  Consider putting a rule into the primary Makefile to do the whole build and clean up anything afterward that should be cleaned up, like we have for rpm.

**Comment by Kunzol** [comment]
> My build completed.

Good !

> Please mention that the target packages go into directory '..' from where you run the command.

Usually the people working with `debuild` know this, but okay, this can be done easily.

> Consider adding some files go .gitignore. After the build 'git status' shows a number of extra files:
> 
> ```
> hanges not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git restore <file>..." to discard changes in working directory)
> 	modified:   dist/debian/changelog
> 	modified:   dist/debian/control
>

A `.gitignore` for changelog does not make sense. The `changelog` file is changed to update the version. Without the `changelog` file the `debuild` process will fail immediately. 

What I did is a kind of "hack" to create the version in the changelog needed for the package. 

In theory the maintainer of the source need to keep the Debian changelog file up to date with every single version (at least every single tag!). 

The control file was modified by you manually changing the "Architecture" line !

> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
> 	debian
> 	dist/debian/.debhelper/
> 	dist/debian/files
> 	dist/debian/singularity-container.debhelper.log
> 	dist/debian/singularity-container.substvars
> 	dist/debian/singularity-container/

These files are temporary files and removed on `dh clean`.  It is not possible to start this clean from the `rules` file. The debuild part after building the binaries will fail, if the binaries are not around anymore. 
I will add a line in the DEBIAN_PACKAGE.md about how to clean up.

File: dist/debian/control (hunk 2 above)
Comment (DrDaveD) [reply dist/debian/control]: The value here though is not `all`, it is `any`.  Does that cause something to break?

**Review by DrDaveD — CHANGES_REQUESTED** [review]
Conflicts need to be resolved by you updating your branch with the latest changes in upstream/master, for example with:
```
git fetch upstream
git checkout master
git merge upstream/master
git push
git checkout - # to switch back to your update-debian-packaging branch
git rebase master
```
You will be asked to resolve conflicts, follow the instructions.  Instead of using rebase `git merge master` should also work but usually rebase is cleaner.  If you don't yet have an upstream remote add it with `git remote add upstream https://github.com/hpcng/singularity.git`.

Very good on the explanation for changes.   At least add "debian" to .gitignore because you want that symlink to stick around.  If the changelog needs to be updated for each tag, then it should be in .github/RELEASE_PROCEDURE.md.  Note that there's another change coming to that file soon due to issue #6019 so you might have another conflict to resolve after changing that.

In summary, the remaining items:
1. Change back the Architecture to "any" unless that causes something to break
2. Add "debian" to .gitignore
3. Add instructions for updating changelog to .github/RELEASE_PROCEDURE.md

**Comment by Kunzol** [comment]
> Conflicts need to be resolved by you updating your branch with the latest changes in upstream/master, for example with:

resolved.

> 
> In summary, the remaining items:
> 
> 1. Change back the Architecture to "any" unless that causes something to break

This is wrong, because "any" means this would be an architecture-independent package, which is not true!

> 2. Add "debian" to .gitignore

Done.

> 3. Add instructions for updating changelog to .github/RELEASE_PROCEDURE.md

I am not sure that this is the right method to go for.

**Comment by Kunzol** [comment]
> > In summary, the remaining items:
> > 
> > 1. Change back the Architecture to "any" unless that causes something to break
> 
> This is wrong, because "any" means this would be an architecture-independent package, which is not true!
> 

Sorry, I mixed up "any" with "all". 
Architecture-independent is marked with "all".

**Comment by DrDaveD** [comment]
I forgot to mention that after you have resolved differences, you need to push the changes to your origin.  If you did a rebase, you need to do `git push -f` to forcibly replace previous commits to the branch when you push it, or it will complain.

> > Conflicts need to be resolved by you updating your branch with the latest changes in upstream/master, for example with:
> 
> resolved.

It's true that it no longer says that there's a conflict, but unfortunately something else is messed up because it is showing that it is trying to re-apply a bunch of changes that have already been committed to master in other PRs.  I'm not entirely sure how to fix that.  Did you remember to do the "git push" from the master branch to update master on your fork on github?  I think that might be one possible cause.  Worse case you might need to recreate the whole branch, cherry-picking your changes from your current branch (or just copying the changes and doing it all with one new commit), and push -f to rebuild the branch.

> > In summary, the remaining items:
> > 
> > 1. Change back the Architecture to "any" unless that causes something to break
> 
> This is wrong, because "any" means this would be an architecture-independent package, which is not true!

No it does not.  "any" is a wildcard matching any architecture.  The architecture is then taken from the build machine.  "all" is architecture-independent.  See the [Debian Architecture documentation](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-architecture).

> > 1. Add "debian" to .gitignore
> 
> Done.

+1

> > 1. Add instructions for updating changelog to .github/RELEASE_PROCEDURE.md
> 
> I am not sure that this is the right method to go for.

It would be better to automate it.  I think the alternative would be to change the build instructions to remove that change after building, with `git checkout HEAD -- dist/debian/changelog`.  In other debian packaging I am familiar with, for other packages, they copy all the source to a scratch area and delete it all after finish.

**Comment by Kunzol** [comment]
> > 1. Add instructions for updating changelog to .github/RELEASE_PROCEDURE.md
> 
> I am not sure that this is the right method to go for.

Would this be something for adding to [Github Workflow](https://docs.github.com/en/actions/reference/events-that-trigger-workflows) to update the changelog file on each release ?

But I am not an expert with Github.

**Comment by Kunzol** [comment]
> I forgot to mention that after you have resolved differences, you need to push the changes to your origin. If you did a rebase, you need to do `git push -f` to forcibly replace previous commits to the branch when you push it, or it will complain.
> 
> > > Conflicts need to be resolved by you updating your branch with the latest changes in upstream/master, for example with:
> > 
> > 
> > resolved.
> 
> It's true that it no longer says that there's a conflict, but unfortunately something else is messed up because it is showing that it is trying to re-apply a bunch of changes that have already been committed to master in other PRs. I'm not entirely sure how to fix that. Did you remember to do the "git push" from the master branch to update master on your fork on github? I think that might be one possible cause. Worse case you might need to recreate the whole branch, cherry-picking your changes from your current branch (or just copying the changes and doing it all with one new commit), and push -f to rebuild the branch.

I expected this to happen. Did I mention, that I am not an expert with Github? :-)

I will try to cleanup the mess, without making it worse.

**Comment by DrDaveD** [comment]
> > > 1. Add instructions for updating changelog to .github/RELEASE_PROCEDURE.md
> > 
> > 
> > I am not sure that this is the right method to go for.
> 
> Would this be something for adding to [Github Workflow](https://docs.github.com/en/actions/reference/events-that-trigger-workflows) to update the changelog file on each release ?

No I don't think that's the way to go.  How about if instead of making debian a symlink that you make the instruction be to do
```
rm -rf debian
cp -r dist/debian .
```
Then the contents of that whole directory won't matter.  The cleanup could also be ignored, since you've got that directory in .gitignore.

**Comment by Kunzol** [comment]
> > > > 1. Add instructions for updating changelog to .github/RELEASE_PROCEDURE.md
> > > 
> > > 
> > > I am not sure that this is the right method to go for.
> > 
> > 
> > Would this be something for adding to [Github Workflow](https://docs.github.com/en/actions/reference/events-that-trigger-workflows) to update the changelog file on each release ?
> 
> No I don't think that's the way to go. How about if instead of making debian a symlink that you make the instruction be to do
> 
> ```
> rm -rf debian
> cp -r dist/debian .
> ```
> 
> Then the contents of that whole directory won't matter. The cleanup could also be ignored, since you've got that directory in .gitignore.

As you like. I don't care. 
I will update the install Instructions.

**Review by DrDaveD — APPROVED** [review]
Awesome.  The only nit I have remaining is that 'dh clean' is no longer needed in the instructions.  Either replace it with 'rm -rf debian' or delete the section on cleaning.  We can get this into 3.8.0.

**Comment by Kunzol** [comment]
> Awesome. The only nit I have remaining is that 'dh clean' is no longer needed in the instructions. Either replace it with 'rm -rf debian' or delete the section on cleaning. We can get this into 3.8.0.

The `dh clean` removes the chosen "BUILDDIR", too. This might not be the same as the standard and might is not in the `.gitignore`. 
I will update the instructions.
